aws_lambda_builders/workflow.py

```diff
@@ -40,25 +40,34 @@ def sanitize(func):
 
     @functools.wraps(func)
     def wrapper(self, *args, **kwargs):
-        valid_paths = []
+        valid_paths = {}
+        invalid_paths = {}
         # NOTE: we need to access binaries to get paths and resolvers, before validating.
         binaries_copy = self.binaries
         for binary, binary_path in binaries_copy.items():
+            invalid_paths[binary] = []
             validator = binary_path.validator
             exec_paths = binary_path.resolver.exec_paths if not binary_path.path_provided else binary_path.binary_path
             for executable_path in exec_paths:
-                valid_path = None
                 try:
                     valid_path = validator.validate(executable_path)
+                    if valid_path:
+                        valid_paths[binary] = valid_path
                 except MisMatchRuntimeError as ex:
                     LOG.debug("Invalid executable for %s at %s", binary, executable_path, exc_info=str(ex))
-                if valid_path:
-                    binary_path.binary_path = valid_path
-                    valid_paths.append(valid_path)
+                    invalid_paths[binary].append(executable_path)
+                if valid_paths.get(binary, None):
+                    binary_path.binary_path = valid_paths[binary]
                     break
         self.binaries = binaries_copy
         if len(self.binaries) != len(valid_paths):
-            raise WorkflowFailedError(workflow_name=self.NAME, action_name=None, reason="Binary validation failed!")
+            validation_failed_binaries = set(self.binaries.keys()).difference(valid_paths.keys())
+            message = ""
+            for validation_failed_binary in validation_failed_binaries:
+                message = "Binary validation failed for {}, Invalid paths : {}".format(
```

Comment:
Not sure this explains to customers what we want. To me, this says "the workflow needed binary X but it has all these invalid paths". But it is not clear to me what invalid paths are and why it matters, or even why we searched that path. Something like "We searched for binary X in these locations Y but did not find X that satisfied our constraints." We will still need a way to tell customers what the constraints are but it gets us closer to being descriptive to the customers so they can understand and fix the problem.

Reply:
Fair enough, with the error message, we can be more explicit.